### DISCUSSION TITLE
Advice with a project development

### DISCUSSION THREAD

**Discussion by santiagorg2401** [USC-ACTLab/crazyswarm#347] [Q&A]
Hi, I hope you're all well,
I've been working with this project for several months so far and I'd like to do something new, for a course project I want to develop a obstacle detection and avoidance algorithm (under the Crazyflie only), for the positioning we have the LPS system with 8 nodes, and for the obstacle detection I will be working with the laser sensor within the flow deck.

The obstacle detection and path correction (please check the picture submitted) modules will be running onboard the Crazyflie (in order to avoid latency problems), I understand that the ROS Server running in the computer sends the desired setpoints to the controller onboard the Crazyflie so it can generate the appropriate signals to its motors, and that I will have to "suppress" the flow deck data in the Kalman State Estimator in order to depend the positioning from the LPS system only (https://forum.bitcraze.io/viewtopic.php?f=19&t=4417&p=20157&hilit=LPS+accuracy#p20157), but, what I do not understand is which files should I modify in order to achieve the objective, also, how to execute onboard the files required for the obstacle detection and avoidance algorithm in the main system loop, and with that I mean, where should those files be located?

Any information is highly appreciated, thank you.

![Screenshot from 2021-03-20 18-12-54](https://user-images.githubusercontent.com/52050852/111888171-17b23c80-89a8-11eb-8bb8-1ba05b13bd62.png)

**Comment by whoenig** [comment]
From what I understand, you want to change the firmware only, to achieve obstacle avoidance of a single Crazyflie with static (or dynamic) obstacles in the environment. I think the existing multi-robot avoidance that @jpreiss recently contributed might be a good starting point: https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/stabilizer.c#L263, https://github.com/bitcraze/crazyflie-firmware/pull/628.

**Comment by santiagorg2401** [comment]
Hi, that is correct, the firmware only, thank you, if I may, just one more question, where is the Crazyflie's main loop?, I mean the file that takes sensor data, the next setpoint, etc.

Thank you for answering, I'd like to contribute the work when I finish it.

**Reply by whoenig** [reply]
https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/stabilizer.c#L237-L299
